tests/test_decoders.py

```diff
@@ -0,0 +1,201 @@
+# -*- coding: utf-8 -*-
+
+from unittest import TestCase
+from unittest.mock import Mock, call, patch
+
+import torch
+
+from steganogan import decoders
+from tests.utils import assert_called_with_tensors
+
+import copy
+
+
+class TestBasicDecoder(TestCase):
+
+    class TestDecoder(decoders.BasicDecoder):
+        def __init__(self):
+            pass
+
+    def setUp(self):
+        self.test_decoder = self.TestDecoder()
+
+    @patch('steganogan.decoders.nn.Conv2d', autospec=True)
+    def test__covn2d(self, conv2d_mock):
+        """Conv2d must be called with given args and kernel_size=3 and padding=1"""
+
+        # run
+        result = self.test_decoder._conv2d(2, 4)
+
+        # asserts
+        assert result == conv2d_mock.return_value
+        conv2d_mock.assert_called_once_with(
+            in_channels=2,
+            out_channels=4,
+            kernel_size=3,
+            padding=1
+        )
+
+    @patch('steganogan.decoders.nn.Sequential')
+    @patch('steganogan.decoders.nn.Conv2d')
+    @patch('steganogan.decoders.nn.BatchNorm2d')
+    def test___init__(self, batchnorm_mock, conv2d_mock, sequential_mock):
+        """Test the init params and that the layers are created correctly"""
+        # setup
+        data_depth = 2
+        hidden_size = 5
+
+        expected_conv_calls = [
+            call(in_channels=3, out_channels=5, kernel_size=3, padding=1),
+            call(in_channels=5, out_channels=5, kernel_size=3, padding=1),
+            call(in_channels=5, out_channels=5, kernel_size=3, padding=1),
+            call(in_channels=5, out_channels=2, kernel_size=3, padding=1),
+        ]
+
+        expected_batch_calls = [call(5), call(5), call(5)]
+
+        # run
+
+        decoder = decoders.BasicDecoder(data_depth, hidden_size)
+
+        # assert
+        assert batchnorm_mock.call_args_list == expected_batch_calls
+        assert conv2d_mock.call_args_list == expected_conv_calls
+
+    def test_upgrade_legacy_without_version(self):
+        """Upgrade legacy must create self._models from conv1, conv2, conv3, conv4"""
+        # setup
+        self.test_decoder.layers = Mock(return_value=torch.Tensor([[5, 6], [7, 8]]))
+
+        expected_version = '1'
```

Comment:
Skip simple variables like this one that have constant values that will be used only once.

use `assert self.test_decoder.version == '1'` instead.